crates/assets/src/asset.rs

```diff
@@ -0,0 +1,895 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+use frame_support::sp_runtime::RuntimeDebug;
+
+/// An error in the asset ID parsing logic.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, RuntimeDebug)]
+pub enum AssetIdError {
+	/// An error in the asset namespace parsing logic.
+	Namespace(NamespaceError),
+	/// An error in the asset reference parsing logic.
+	Reference(ReferenceError),
+	/// An error in the asset identifier parsing logic.
+	Identifier(IdentifierError),
+	/// A generic error not belonging to any of the other categories.
+	InvalidFormat,
+}
+
+/// An error in the asset namespace parsing logic.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, RuntimeDebug)]
+pub enum NamespaceError {
+	/// Namespace too long.
+	TooLong,
+	/// Namespace too short.
+	TooShort,
+	/// A generic error not belonging to any of the other categories.
+	InvalidFormat,
+}
+
+/// An error in the asset reference parsing logic.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, RuntimeDebug)]
+pub enum ReferenceError {
+	/// Reference too long.
+	TooLong,
+	/// Reference too short.
+	TooShort,
+	/// A generic error not belonging to any of the other categories.
+	InvalidFormat,
+}
+
+/// An error in the asset identifier parsing logic.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, RuntimeDebug)]
+pub enum IdentifierError {
+	/// Identifier too long.
+	TooLong,
+	/// Identifier too short.
+	TooShort,
+	/// A generic error not belonging to any of the other categories.
+	InvalidFormat,
+}
+
+impl From<NamespaceError> for AssetIdError {
+	fn from(err: NamespaceError) -> Self {
+		Self::Namespace(err)
+	}
+}
+
+impl From<ReferenceError> for AssetIdError {
+	fn from(err: ReferenceError) -> Self {
+		Self::Reference(err)
+	}
+}
+
+impl From<IdentifierError> for AssetIdError {
+	fn from(err: IdentifierError) -> Self {
+		Self::Identifier(err)
+	}
+}
+
+// Exported types. This will always only re-export the latest version by
+// default.
+pub use v1::*;
+
+pub mod v1 {
+	use super::{AssetIdError, IdentifierError, NamespaceError, ReferenceError};
+
+	use codec::{alloc::string::ToString, Decode, Encode, MaxEncodedLen};
+	use scale_info::TypeInfo;
+
+	use core::{format_args, str};
+
+	use frame_support::{sp_runtime::RuntimeDebug, traits::ConstU32, BoundedVec};
+	use sp_core::U256;
+	use sp_std::{fmt::Display, vec::Vec};
+
+	pub const MINIMUM_ASSET_ID_LENGTH: usize = MINIMUM_NAMESPACE_LENGTH + 1 + MINIMUM_REFERENCE_LENGTH;
+	pub const MAXIMUM_ASSET_ID_LENGTH: usize =
+		MAXIMUM_NAMESPACE_LENGTH + 1 + MAXIMUM_REFERENCE_LENGTH + 1 + MAXIMUM_IDENTIFIER_LENGTH;
+
+	pub const MINIMUM_NAMESPACE_LENGTH: usize = 3;
+	pub const MAXIMUM_NAMESPACE_LENGTH: usize = 8;
+	const MAXIMUM_NAMESPACE_LENGTH_U32: u32 = MAXIMUM_NAMESPACE_LENGTH as u32;
```

Comment:
Yeah I also hate it, but the thing is that `BoundedVec` use `u32` for sizes, and everything else in Rust uses `usize`. Instead of casting everytime, I just do it once in a constant and be done with it.

Reply:
@weichweich do you have a better solution?

Reply:
No not really. 😢
Any reason for the cast direction? usize as u32 and not the other way around? AFAIK it doesn't matter since we are targeting wasm where usize is u32. But on most of our CPUs usize would be u64 so the cast could overflow. I can see that `8` will probably not cause issues.

Reply:
Yeah but then we would have to cast each time we want to use the value, which is what was annoying me.